extmod/machine_wdt.h

```diff
@@ -0,0 +1,48 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2023 Damien P. George
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#ifndef MICROPY_INCLUDED_EXTMOD_MACHINE_WDT_H
+#define MICROPY_INCLUDED_EXTMOD_MACHINE_WDT_H
+
+#include "py/obj.h"
+
+// A port must provide this type, but it's otherwise opaque.
+typedef struct _machine_wdt_obj_t machine_wdt_obj_t;
+
+// This WDT class is implemented by machine_wdt.c.
+extern const mp_obj_type_t machine_wdt_type;
+
+// A port must provide implementations of these low-level WDT functions, either as global
+// linker symbols, or included directly if MICROPY_PY_MACHINE_WDT_INCLUDEFILE is defined.
+#ifndef MICROPY_PY_MACHINE_WDT_INCLUDEFILE
+machine_wdt_obj_t *mp_machine_wdt_make_new_instance(mp_int_t id, mp_int_t timeout_ms);
```

Comment:
What about declaring these as STATIC if `MICROPY_PY_MACHINE_WDT_INCLUDEFILE` is set, instead of not declaring them?

The compilation benefit is pretty minor (compiler can check the definitions in the include files against the declaration), but it more clearly hints to the implementer what's expected in that case.

Reply:
Yeah, it would be nice to expose these declarations, so the compiler can check the types.

But if you make these static then it gives errors:
```
CC modmachine.c
In file included from modmachine.c:46:0:
../../extmod/machine_wdt.h:41:27: error: 'mp_machine_wdt_make_new_instance' declared 'static' but never defined [-Werror=unused-function]
 STATIC machine_wdt_obj_t *mp_machine_wdt_make_new_instance(mp_int_t id, mp_int_t timeout_ms);
                           ^
../../extmod/machine_wdt.h:42:13: error: 'mp_machine_wdt_feed' declared 'static' but never defined [-Werror=unused-function]
 STATIC void mp_machine_wdt_feed(machine_wdt_obj_t *self);
             ^
cc1: all warnings being treated as errors
```

What you'd need to fix this is a special macro that's set by the includer (`extmod/machine_wdt.c`) that only makes them static when included and used.

Reply:
I made an attempt to fix this.  See the most recent commit in this PR.

Reply:
Ah, right - I missed that other headers include `machine_wdt.h`. This particular section declares the private interface to `machine_wdt` though, doesn't it? i.e. modmachine.c and other source files don't ever need these declarations.

I think the conventional approach here would be to move the "private" declarations into `machine_wdt_priv.h` or something, and have this "private" header only included from `extmod/machine_wdt.c`  and any port-specific `machine_wdt.c`.

However there is more of a tradeoff to doing that compared to my original suggestion though, so I'm less sure it's worthwhile.

Reply:
A couple of alternative approaches:
- keep private things in `extmod/machine_wdt.h` and move the public declarations (really just the `machine_wdt_type`) to a common `extmod/modmachine.h` file
- add a wrapper macro around these private declarations, eg `#ifdef MICROPY_MACHINE_WDT_NEED_PRIVATE_DECLS`, and that must be enabled by any user of the file that needs the declarations

Reply:
Oops, our replies crossed. Doing it that way looks like it works pretty well, although I think it's a bit hard to understand just by reading it (if you include `machine_wdt.h` from another source file then you're going to see a bunch of phony declarations.).

I think I slightly prefer keeping it the way it was, or going all the way to split the header into a public and private interface to make the intention really explicit (unless there's another constraint here that I don't see.)

Reply:
Crossed again. :laughing: 

>  keep private things in extmod/machine_wdt.h and move the public declarations (really just the machine_wdt_type) to a common extmod/modmachine.h file

If this is all that's needed (and the approach scales across the other machine peripherals) then I reckon that's the cleanest way to factor it.

Reply:
Another approach: we could force all ports to use the include-file approach (they all currently do anyway), then the static declarations will always be static and they can just be put directly at the top of `machine_wdt.c`, before including the include-file.

Then the public things can go in `extmod/modmachine.h`.

That way we don't have a whole bunch of `extmod/machine_XXX.h` headers, and I'm always in favour of reducing the number of files lying around (although there are some headers like `extmod/machine_i2c.h` that would be best to remain separate from `extmod/modmachine.h`).

If a port wants to provide non-static versions of the implementation functions, then its include-file can be a simple set of static wrapper functions that call out to the non-static versions.  In other words, making the include-file approach mandatory we make it a bit simpler to do this approach (which all ports use) and a bit more complicated to do the other approach which is currently not even used (namely, when the include-file is not set).

Reply:
> Another approach: we could force all ports to use the include-file approach (they all currently do anyway

I was considering asking whether there were ports that don't or couldn't use the static approach. Fully agree with your reasoning about it, that sounds like a good way forward and it simplifies a bunch of things.

Reply:
OK, I applied the following:
- made a new `extmod/modmachine.h` header with only the public WDT stuff in it
- removed `extmod/machine_wdt.h`
- put static declarations of port-provided functions at the top of `extmod/machine_wdt.c`
- made the include-file mandatory

Reply:
Changes look good to me, thanks!